src/lib/comms/sol-coap.c

```diff
@@ -763,6 +731,8 @@ prepare_buffer(struct outgoing *outgoing, struct sol_buffer *buffer)
         payload->buf.used - old_offset);
     SOL_INT_CHECK_GOTO(r, < 0, error);
 
+    buffer->used = new_size;
```

Comment:
Just pushed an updated version, with those fixes. 

@glima, just added a few more fixes too, see if it makes sense.